mpqp/qasm/open_qasm_2_and_3.py

```diff
@@ -740,3 +747,321 @@ def remove_include(qasm_code: str) -> str:
+            else:
+                instructions_code += f"measure {q} -> {c};\n"
+    elif instr_name == "u3":
+        instructions_code += re.sub(r"\bu3\b", "U", instr) + ";\n"
+    elif instr_name == "cp":
+        instructions_code += re.sub(r"\bcp\b", "cu1", instr) + ";\n"
+
+    elif instr_name in std_gates_3_to_2_map:
+        converted_instr_name = std_gates_3_to_2_map[instr_name]
+        instructions_code += (
+            re.sub(r"\b" + instr_name + r"\b", converted_instr_name, instr) + ";\n"
+        )
+    elif instr_name in std_gates_3:
+        m = re.match(r"\s*(.*)", instr)
+        if m:
+            instructions_code += m.group(1) + ";\n"
+    elif instr_name == "gate":
+        m = re.match(
+            r"\s*gate\s+(\w+)\s*(\(([^)]*)\))?\s*([\w\s,]*)\s*{\s*([^}]*)\s*}",
+            instr,
+            re.DOTALL,
+        )
+        if m:
+            gate_name = m.group(1)
+            params = f"({m.group(3)})" if m.group(3) else ""
+            qubits = m.group(4).strip()
+            body = m.group(5)
+
+            defined_gates.add(gate_name)
+            g_string = f"gate {gate_name}{params} {qubits} {{\n"
+            g_instructions = filter(
+                lambda i: not re.fullmatch(r"\s*", i), body.split(";")
+            )
+            for instruction in g_instructions:
+                instruction = instruction.strip()
+                i_code, h_code = convert_instruction_3_to_2(
+                    instruction,
+                    included_instr,
+                    included_tree_current,
+                    defined_gates,
+                    path_to_main,
+                )
+                g_string += " " * 4 + i_code  # Add indentation to body instructions
+                header_code += h_code
+
+            # Finalize the gate and add it to instructions_code
+            instructions_code += g_string + "}\n"
+
+    elif re.match(r"\s*if\s*\(.*?\)\s*.+", instr):
+        m = re.match(r"(\s*if\s*\(.*?\))\s*(.+)", instr)
+        if m:
+            if_statement = m.group(1)
+            nested_instr = m.group(2)
+            i_code, h_code = convert_instruction_3_to_2(
+                nested_instr,
+                included_instr,
+                included_tree_current,
+                defined_gates,
+                path_to_main,
+            )
+            instructions_code += if_statement + i_code + ";\n"
+            header_code += h_code
+
+    else:
+        gate = instr.split()[0]
+        if gate == "ctrl":
+            gate = instr.split()[2]
+        if gate not in defined_gates:
+            raise ValueError(f"Gates undefined at the time of usage: {gate}")
+        m = re.match(r"\s*(.*)", instr)
+        if m:
+            instructions_code += m.group(1) + ";\n"
+
+    return instructions_code, header_code
+
+
+@typechecked
+def open_qasm_3_to_2(
+    code: str,
+    included_tree_current_node: Node = Node("initial_code"),
```

Comment:
In python setting default params to objects is not recommended because this object would be the same (it would be the same reference to a mutable object) to all function call, so it you modify it, you would modify it for all future function call. Instead default it to `None` and in the body of the function, add a `if a is None: a = ...`
For an example of messed up default parameter initialization, you can for example run
```py
>>> def test(a = []): a.append(1); print(a)
>>> test()
[1]
>>> test()
[1, 1]
```

Reply:
done